File: Part 02 - Regression/Section 5 - Multiple Linear Regression/Multiple_Linear_Regresion.py

# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd

# Importing the dataset
dataset = pd.read_csv('50_Startups.csv')
X = dataset.iloc[:, :-1].values
y = dataset.iloc[:, 4].values

# Encoding categorical data
from sklearn.preprocessing import LabelEncoder, OneHotEncoder
labelencoder_X = LabelEncoder()
X[:, 3] = labelencoder_X.fit_transform(X[:, 3])
onehotencoder = OneHotEncoder(categorical_features = [3])
X = onehotencoder.fit_transform(X).toarray()

# Avoiding the Dummy Variable Trap
X = X[:, 1:]
                
# Splitting the dataset into the Training set and Test set
from sklearn.cross_validation import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

# Fitting Multiple Linear Regression to the Training set
from sklearn.linear_model import LinearRegression
regressor = LinearRegression()
regressor.fit(X_train, y_train)

# Predicting the Test set results
y_pred = regressor.predict(X_test)

#Building the optimal model 
import statsmodels.formula.api as sm
#Now we add the first column to the matrix X, belonging to the b0 
#constant of our Multiple Linear Regression Model. 
X = np.append(arr = np.ones((50, 1)).astype(int), values = X, axis = 1)
#A matrix containing all the independent variables we have in our data. 
X_opt = X[:, [0,1,2,3,4,5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary() 
#Now we remove one of the variables that gives us a p_value > significance
#value which we set previously as 0.05
X_opt = X[:, [0,1,2,4,5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary() 

X_opt = X[:, [0,1,4,5]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

X_opt = X[:, [0,1,4]]
regressor_OLS = sm.OLS(endog=y, exog=X_opt).fit()
regressor_OLS.summary()

File: Part 02 - Regression/Section 5 - Multiple Linear Regression/Multiple_Linear_Regresion.R
# Data Preprocessing Template

# Importing the dataset

dataset = read.csv('50_Startups.csv')
dataset$State = factor(dataset$State, 
                       levels = c('New York', 'California', 'Florida'),
                       labels = c(1,2,3))


# Splitting the dataset into the Training set and Test set
# install.packages('caTools')
library(caTools)
set.seed(123)
split = sample.split(dataset$Profit, SplitRatio = 0.8)
training_set = subset(dataset, split == TRUE)
test_set = subset(dataset, split == FALSE)

#Feature Scaling won't be needed since this will be handled by the function
#that is used to fit multiple linear regression to our training set. 

regressor = lm(formula = Profit  ~ R.D.Spend + Administration
               + Marketing.Spend + State, data = training_set)
#This above statement can be written also as follows:
# regressor = lm(formula = Profit  ~ .,data=training_set)
summary(regressor)
y_pred = predict(regressor, newdata = test_set)

#Find the best optimal model for our problem with Backward Elimination. 
#that is used to fit multiple linear regression to our training set. 
#--------------------------------------------------------------------------#
#We input now the independent variables with + so when we implement the step 5 of backward
#elimination we can remove the variable that is not statistically significant. 

regressor = lm(formula = Profit  ~ R.D.Spend + Administration +  Marketing.Spend + State,
               data = dataset)
#--------------------------------------------------------------------------#
#We are taking the whole dataset so we can see which variables are statistically 
#significant and which not. 
summary(regressor)

#1 - Removing the variable that gives a p_value above the SL. 

regressor = lm(formula = Profit  ~ R.D.Spend + Administration +  Marketing.Spend,
               data = dataset)
#--------------------------------------------------------------------------#
#We are taking the whole dataset so we can see which variables are statistically 
#significant and which not. 
summary(regressor)

#2 - Removing the second variable that gives a p_value above the SL. 

regressor = lm(formula = Profit  ~ R.D.Spend + Marketing.Spend,
               data = dataset)
#--------------------------------------------------------------------------#
#We are taking the whole dataset so we can see which variables are statistically 
#significant and which not. 
summary(regressor)

#3 - Removing the second variable that gives a p_value above the SL. 
regressor = lm(formula = Profit  ~ R.D.Spend,
               data = dataset)
#--------------------------------------------------------------------------#
#We are taking the whole dataset so we can see which variables are statistically 
#significant and which not. 
summary(regressor)

#Either you eliminate the Marketing.Spend variable or keep it for your model is ok!.



